pkg/controllers/selection/preferences.go

```diff
@@ -64,18 +64,23 @@ func (p *Preferences) Relax(ctx context.Context, pod *v1.Pod) {
 	cachedSpec := spec.(v1.PodSpec)
 	pod.Spec.Affinity = cachedSpec.Affinity
 	pod.Spec.Tolerations = cachedSpec.Tolerations
-	if relaxed := p.relax(ctx, pod); relaxed {
+	if relaxed := p.relax(ctx, pod, addPreferNoScheduleToleration); relaxed {
 		p.cache.SetDefault(string(pod.UID), pod.Spec)
 	}
 }
 
-func (p *Preferences) relax(ctx context.Context, pod *v1.Pod) bool {
+func (p *Preferences) relax(ctx context.Context, pod *v1.Pod, addPreferNoScheduleToleration bool) bool {
 	for _, relaxFunc := range []func(*v1.Pod) *string{
 		func(pod *v1.Pod) *string { return p.removePreferredPodAffinityTerm(pod) },
 		func(pod *v1.Pod) *string { return p.removePreferredPodAntiAffinityTerm(pod) },
 		func(pod *v1.Pod) *string { return p.removePreferredNodeAffinityTerm(pod) },
 		func(pod *v1.Pod) *string { return p.removeRequiredNodeAffinityTerm(pod) },
-		func(pod *v1.Pod) *string { return p.toleratePreferNoScheduleTaints(pod) },
+		func(pod *v1.Pod) *string {
+			if addPreferNoScheduleToleration {
+				return p.toleratePreferNoScheduleTaints(pod)
```

Comment:
Seems like this check out encapsulate `addPreferNoScheduleToleration`, rather than having this logic in the outer for loop.

Reply:
Turned it into a generic options struct, seems to make more sense.